File: src/renderer/gpu/colors.ts
// Per-row color storage buffer — one vec4<f32> (RGBA, 16B) per GPU row,
// indexed in the shader by `segment.rowFlags & 0xffff`. Host writes hex
// strings; we normalize to [0,1] f32.
// Max active rows. Bounded by the 16-bit row index in the GPU segment's row_flags
// (≤ 65535); must match native segments.zig MAX_ROWS. The color buffer below is one
// vec4 per row = MAX_ROWS×16 B ≈ 1 MB, allocated once (negligible GPU memory); only
// the used prefix is uploaded per write (see writeRowColors).
export const MAX_ROWS = 65535;
export const ROW_COLOR_STRIDE = 16;

export function createColorBuffer(device: GPUDevice): GPUBuffer {
  return device.createBuffer({
    size: MAX_ROWS * ROW_COLOR_STRIDE,
    usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_DST,
  });
}

export function writeRowColors(
  device: GPUDevice,
  buf: GPUBuffer,
  entries: { row: number; color: string }[],
): void {
  // Upload only the used prefix [0 .. maxRow]. The GPU buffer is sized for MAX_ROWS,
  // but allocating that full array each call (a ~1 MB Float32Array at the 65535 cap)
  // on every cosmetic change is needless GC. Rows past the prefix keep stale-but-
  // unreferenced colors (no segment's row index points at them).
  let maxRow = -1;
  for (const { row } of entries) {
    if (row < 0 || row >= MAX_ROWS) throw new Error(`row ${row} out of range (max ${MAX_ROWS})`);
    if (row > maxRow) maxRow = row;
  }
  if (maxRow < 0) return;
  const packed = new Float32Array((maxRow + 1) * 4);
  for (const { row, color } of entries) {
    const [r, g, b, a] = hexToRgba(color);
    const off = row * 4;
    packed[off + 0] = r;
    packed[off + 1] = g;
    packed[off + 2] = b;
    packed[off + 3] = a;
  }
  device.queue.writeBuffer(buf, 0, packed);
}

function hexToRgba(hex: string): [number, number, number, number] {
  const h = hex.replace("#", "");
  return [
    parseInt(h.slice(0, 2), 16) / 255,
    parseInt(h.slice(2, 4), 16) / 255,
    parseInt(h.slice(4, 6), 16) / 255,
    1,
  ];
}

File: crates/riptide-render/src/colors.rs
//! Per-row color storage buffer (port of `gpu/colors.ts`): MAX_ROWS vec4<f32>
//! colors, rewritten whenever the active-signal colors change.

use riptide_contract::gpu::MAX_ROWS;

pub struct ColorBuffer {
    pub buf: wgpu::Buffer,
}

impl ColorBuffer {
    pub fn new(device: &wgpu::Device) -> Self {
        let buf = device.create_buffer(&wgpu::BufferDescriptor {
            label: Some("row-colors"),
            size: (MAX_ROWS * 16) as u64, // vec4<f32> per row
            usage: wgpu::BufferUsages::STORAGE
                | wgpu::BufferUsages::COPY_DST
                // COPY_SRC so tests/captures can read the palette back.
                | wgpu::BufferUsages::COPY_SRC,
            mapped_at_creation: false,
        });
        Self { buf }
    }

    /// Writes packed-rgba row colors (0xAABBGGRR little-endian, as everywhere
    /// else) as normalized vec4<f32>s. `colors[row]`, missing rows → 0.
    pub fn write(&self, queue: &wgpu::Queue, colors: &[u32]) {
        let mut data = [[0.0f32; 4]; MAX_ROWS];
        for (i, &c) in colors.iter().take(MAX_ROWS).enumerate() {
            let b = c.to_le_bytes();
            data[i] = [
                b[0] as f32 / 255.0,
                b[1] as f32 / 255.0,
                b[2] as f32 / 255.0,
                b[3] as f32 / 255.0,
            ];
        }
        queue.write_buffer(&self.buf, 0, bytemuck::cast_slice(&data));
    }
}
